File: lab5/haskell/totient.hs
import System.CPUTime
import Text.Printf
import Data.List (group)


coprime :: Int -> [Int]
coprime n = [k | k <- [1..n], gcd n k == 1]

totient :: Int -> Int
totient n = length $ coprime n


factorize :: Int -> Int -> [Int]
factorize 1 _ = []
factorize n divisor
    | n `mod` divisor == 0 = divisor : factorize (n `div` divisor) divisor
    | otherwise            = factorize n (divisor + 1)

primeFactors :: Int -> [Int]
primeFactors n
    | n <= 1    = []
    | otherwise = factorize n 2

totient2 :: Int -> Int
totient2 n =
    let factors = primeFactors n
        uniqueFactors = map head $ group factors
    in round $ fromIntegral n * product [1 - 1 / fromIntegral p | p <- uniqueFactors]


timeIt :: IO a -> IO Double
timeIt action = do
    start <- getCPUTime
    _ <- action
    end <- getCPUTime
    let diff = fromIntegral (end - start) / (10^12)
    return diff


main :: IO ()
main = do
    let n = 39916801

    time1 <- timeIt $ printf "totient %d = %d\n" n (totient n)
    printf "Time for totient %d : %0.3fs\n" n time1

    printf "\n"

    time2 <- timeIt $ printf "totient2 %d = %d\n" n (totient2 n)
    printf "Time for totient2 %d : %0.3fs\n" n time2

File: lab5/prolog/totient.pl
gcd(A, 0, A).
gcd(A, B, GCD) :-
    B > 0,
    C is A mod B,
    gcd(B, C, GCD).

coprime(X, Y) :-
    gcd(X, Y, GCD),
    GCD =:= 1.

totientHelper(_, 1, T, T) :- !.
totientHelper(N, M, Acc, T) :-
    M > 1,
    coprime(N, M),
    NewAcc is Acc + 1,
    NewM is M - 1,
    totientHelper(N, NewM, NewAcc, T).
totientHelper(N, M, Acc, T) :-
    M > 1,
    \+ coprime(N, M),
    NewM is M - 1,
    totientHelper(N, NewM, Acc, T).

totient(1, 1).
totient(N, T) :-
    N > 1,
    totientHelper(N, N, 1, T).


main :-
    N = 7919,
    totient(N, T),
    format("Euler's Totient function for N = ~d is ~d~n", [N, T]).

:- main.
